test/e2e/storage/vsphere_volume_perf.go

```diff
@@ -0,0 +1,264 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/api/core/v1"
+	storageV1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	vsphere "k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+/* This test calculates latency numbers for volume lifecycle operations
+1. Create 4 type of storage classes
+2. Read the total number of volumes to be created and volumes per pod
+3. Create total PVCs (number of volumes)
+4. Create Pods with attached volumes per pod
+5. Verify access to the volumes
+6. Delete pods and wait for volumes to detach
+7. Delete the PVCs
+*/
+const (
+	NodeLabelKey              = "vsphere_e2e_label"
+	SCSIUnitsAvailablePerNode = 55
+	NumOps                    = 4 //Number of volume operations whose latency is to be calculated
+)
+
+// NodeSelector holds
+type NodeSelector struct {
+	labelKey   string
+	labelValue string
+}
+
+var _ = SIGDescribe("vcp-performance [Feature:vsphere]", func() {
+	f := framework.NewDefaultFramework("vcp-performance")
+
+	var (
+		client           clientset.Interface
+		namespace        string
+		nodeSelectorList []*NodeSelector
+		volumeCount      int
+		volumesPerPod    int
+		iterations       int
+	)
+
+	BeforeEach(func() {
+		var err error
+		framework.SkipUnlessProviderIs("vsphere")
+		client = f.ClientSet
+		namespace = f.Namespace.Name
+
+		Expect(os.Getenv("VSPHERE_SPBM_GOLD_POLICY")).NotTo(BeEmpty(), "env VSPHERE_SPBM_GOLD_POLICY is not set")
+		Expect(os.Getenv("VSPHERE_DATASTORE")).NotTo(BeEmpty(), "env VSPHERE_DATASTORE is not set")
+		Expect(os.Getenv("VCP_PERF_VOLUME_PER_POD")).NotTo(BeEmpty(), "env VCP_PERF_VOLUME_PER_POD is not set")
+		Expect(os.Getenv("VCP_PERF_ITERATIONS")).NotTo(BeEmpty(), "env VCP_PERF_ITERATIONS is not set")
+		Expect(os.Getenv("VCP_PERF_VOLUME_COUNT")).NotTo(BeEmpty(), "env VCP_PERF_VOLUME_COUNT is not set")
```

Comment:
Why are you getting each env var twice? Once here and once below in the Atio call.
Also,  these env vars are retrieved for every `It` test. It is possible to define these vars once in the `SIGDescribe` main code rather than in a `BeforeEach`.

Reply:
Corrected them to read it only once.
Since the assertions are used, it cannot be moved to SIGDescribe main code.
This code is supposed to moved to a BeforeSuite construct for all vsphere tests since most of these vars are common.
Tracking through https://github.com/vmware/kubernetes/issues/353

Reply:
ack